service/history/api/recordchildworkflowcompleted/api.go

```diff
@@ -38,21 +40,51 @@ import (
 	"go.temporal.io/server/service/history/workflow"
 )
 
+// maxResetRedirectCount is the number of times we follow the reset run ID to forward the request to the new parent.
+// This redirection happens only when a workflow is reset and in most cases it's 1 or 2 hops.
+// maxResetRedirectCount prevents us from following long chain of resets (or some circular loop in redirects).
+const maxResetRedirectCount = 100
+
 func Invoke(
 	ctx context.Context,
 	request *historyservice.RecordChildExecutionCompletedRequest,
-	shard shard.Context,
+	shardContext shard.Context,
 	workflowConsistencyChecker api.WorkflowConsistencyChecker,
 ) (resp *historyservice.RecordChildExecutionCompletedResponse, retError error) {
-	_, err := api.GetActiveNamespace(shard, namespace.ID(request.GetNamespaceId()))
+	_, err := api.GetActiveNamespace(shardContext, namespace.ID(request.GetNamespaceId()))
 	if err != nil {
 		return nil, err
 	}
 
+	redirectCount := 0
+	for {
+		resetRunID, err := recordChildWorkflowCompleted(ctx, request, shardContext, workflowConsistencyChecker)
+		if errors.Is(err, consts.ErrWorkflowCompleted) {
+			if resetRunID != "" && redirectCount < maxResetRedirectCount {
```

Comment:
I think we return an error (other than not found) when max redirect count is exceeded. 

Otherwise theoretically we could silently drop the child result and parent after reset could get stuck.

Reply:
Good point. Done.